torch/csrc/autograd/record_function.cpp

```diff
@@ -11,177 +11,189 @@ namespace profiler {
 
 namespace {
 
-float sample_zero_one() {
-  static thread_local auto gen =
-      torch::make_unique<std::mt19937>(std::random_device()());
-  std::uniform_real_distribution<float> dist(0.0, 1.0);
-  return dist(*gen);
+// Used to generate unique callback handles
+CallbackHandle next_unique_callback_handle() {
+  static std::atomic<uint64_t> unique_id {0};
+  return CallbackHandle(++unique_id);
 }
 
+// Thread local vector of callbacks, holds pairs (callbacks, unique_id);
+// must be sorted in increasing handles order
+thread_local RecordFunctionCallbacks sorted_tls_callbacks_;
+
 class CallbackManager {
  public:
-  void pushCallback(
-      std::function<bool(const RecordFunction&)> start,
-      std::function<void(const RecordFunction&)> end,
-      bool needs_inputs,
-      double sampling_prob,
-      std::unordered_set<RecordScope, std::hash<RecordScope>> scopes) {
-    callbacks_.emplace_back(
-      std::move(start),
-      std::move(end),
-      needs_inputs,
-      sampling_prob,
-      std::move(scopes)
-    );
-    recomputeFlags();
-
-    // make sure we mark the change in callbacks
-    ++callbacks_version_;
+  CallbackHandle addThreadLocalCallback(RecordFunctionCallback cb) {
+    // note: monotonically increasing callbacks_unique_id keeps
+    // sorted_tls_callbacks_ sorted
+    auto handle = next_unique_callback_handle();
+    sorted_tls_callbacks_.emplace_back(std::move(cb), handle);
+    return handle;
+  }
+
+  CallbackHandle addGlobalCallback(RecordFunctionCallback cb) {
+    auto handle = next_unique_callback_handle();
+    sorted_global_callbacks_.emplace_back(std::move(cb), handle);
+    return handle;
   }
 
-  void popCallback() {
-    if (callbacks_.empty()) {
-      throw std::runtime_error("Empty callbacks stack");
+  void removeCallback(CallbackHandle handle) {
+    bool found = false;
+    auto find_and_remove = [handle, &found](RecordFunctionCallbacks& cbs) {
+      auto it = std::find_if(
+        cbs.begin(), cbs.end(),
+        [handle](
+            const std::pair<
+                RecordFunctionCallback,
+                CallbackHandle>& el) {
+          return el.second == handle;
+        });
+      if (it != cbs.end()) {
+        // keeps it sorted
+        cbs.erase(it);
+        found = true;
```

Comment:
nit: might be a bit more readable if you were just returning a bool instead of modifying `found`